RSS Feed Subscription Integration

flamingquaks opened the pull request:
# RSS Subscription Integration
This Feature introduces the ability for a RAG workspace to subscribe to an RSS Feed, check for new posts daily and send new posts for Website Crawling for the RAG Workspace. This can showcase simple automation for consuming data for RAG / Semantic Search.

![Screenshot 2023-11-17 at 16 16 28](https://github.com/aws-samples/aws-genai-llm-chatbot/assets/40297962/72024687-24c7-4b12-9604-78270a27d298)


## How it Works
Users can subscribe to a RSS Feed by navigating to a desired RAG workspace and navigating to the "RSS Feeds" tab on the workspace documents table. 

Clicking the "Subscribe to RSS Feed" button on the table or by clicking "Add RSS Subscription" in the "Add Data" drop down will navigate the user to the RAG Workspace's "Add Data" view, with the RSS Feeds Tab selected. 
The user can paste a link to an RSS Feed and provide a title for the feed. 

![Screenshot 2023-11-22 at 18 21 03](https://github.com/aws-samples/aws-genai-llm-chatbot/assets/40297962/7b710218-6f9d-4a00-900c-0cd1ad20ae04)


Once the user successfully subscribes to the RSS Feed, the feed will be visible in the "RSS Feeds" tab of the RAG Workspace documents table. 
The RSS Feed is then sent to be checked for posts, which are then added to the RSS Feed DynamoDB table.

An independent scheduled lambda checks the RSS Feed table for posts that are in a pending status and sends up to 10 pending posts for to be Website Crawled into the RAG Workspace. The RSS Feed Check & Post Crawling were designed to be decoupled to provide a fixed rate-limit for the number of websites being crawled. Additionally, the architecture leverages an EventBridge schedule for each post with a flexible timeframe for execution. This can also help distribute work over time if the user has subscribed to a high-volume of feeds or feeds that have a high-volume of posts. Since the posts are ingested using the existing crawl websites functionality, the crawled blog posts will show under the websites once they are consumed. 

Additionally, a user can click on a RSS Feed subscription from the documents table on the RAG workspace, which takes the user to a RSS Feed details page. The page gives the details of the RSS Feed subscription and the posts that have been found from the feed. It also provides the user with the ability to enable/disable a subscription. 

![Screenshot 2023-11-22 at 18 19 11](https://github.com/aws-samples/aws-genai-llm-chatbot/assets/40297962/153dac5a-3bc8-4d84-b20f-bbaa50c681b2)
![Screenshot 2023-11-22 at 18 20 35](https://github.com/aws-samples/aws-genai-llm-chatbot/assets/40297962/adfdb8cd-60bd-4d85-a0fd-5227096f8ad4)


Finally, if a user deletes a RAG workspace, the RSS Subscription's EventBridge schedule is deleted and any RSS Feed or Post data relating to the workspace is removed from the Documents DynamoDB table.


![Nov-17-2023 16-40-20](https://github.com/aws-samples/aws-genai-llm-chatbot/assets/40297962/60bf66c2-f71c-431a-b735-5898960f4ba3)


## High-Level Technical Changes/Additions
The goal with changes was to leverage as much of the existing design patterns/structures. 
* 3 New lambdas are added. 
* 1. The first lambda runs once daily by EventBridge rule. This lambda pulls all RSS Feeds that are currently enabled across any workspace. It loops through each and Invokes an RSS Ingestor Lambda function.
* 1. The second lambda is triggered by the previous lambda. This lambda will parse the RSS feed for the provided `workspace_id`/`feed_id`. The lambda will add newly detected posts to the documents table as `rsspost` type documents. The documents have a status of `pending`. If a post already exists, the lambda will not overwrite it to preserve the original data and ensure process RSS feed posts aren't re-processed.
* 1. The third lambda runs every 10 minutes. This lambda queries the documents table for documents with a status of `pending` and a document type of `rsspost`. It then submits each post for website crawling. The website crawling will be associated with the workspace the RSS feed/posts exist.

The design of the three lambda architecture was based on a few factors:

* The first lambda, which triggers one or more of the second lambda was designed to be the single RSS post check trigger, keeping scheduling simple. The decision to have each RSS feed execute feed crawling in an isolated lambda was to ensure scalability. While feed parsing is relatively quick, if users subscribe to a high volume of feeds, an execution time out could occur. 
* With the second lambda executing feed parsing on one feed per execution, it can help mitigate failures due DDB capacity issues (in high volume scenarios) by allowing smaller retries w/ exponential back off. Additionally, debugging and tracing is easier with isolated feed parsing executions.
* The third lambda is designed to ensure the Website Crawling SageMaker Inference endpoint isn't overwhelmed. Leveraging DDB Documents Table as a processing queue, it grabs a max of 10 websites every 10 minutes. 



This PR is relating to #226.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

bigadsoleiman commented on package.json:
```diff
@@ -38,7 +38,8 @@
     "@aws-cdk/aws-cognito-identitypool-alpha": "^2.104.0-alpha.0",
-    "aws-cdk-lib": "2.104.0",
+    "@aws-solutions-constructs/aws-eventbridge-lambda": "^2.46.0",
+    "aws-cdk-lib": "2.105.0",
```
We want to be consistent on versioning and we usually bump on dedicated PR after testing deployments of the all infra with all components before doing so

Any specific reason to bump to 2.105 here? can this be left at 2.104 for this PR?

flamingquaks replied:
@bigadsoleiman I agree with you about version changes. 
I introduced the AWS solutions construct which had peer dependencies on v2.105. 
Alternatively, I can remove the construct and simply create an event bridge rule manually. There's little effort to make that change and I can revert the version bump.

massi-ang replied:
If you need that specific version because the feature you are using is only available there, ensure you bump the versions of all the CDK packages.

bigadsoleiman commented:
Great contribution @flamingquaks! looking into it

flamingquaks commented:
@bigadsoleiman thanks!
I'm hoping to have finally tweaks done today. I had some over-engineering introduced which I'm yanking out. The final push should be today and I'll take the PR out of draft.

flamingquaks commented:
I've moved the PR back to ready!

From the original PR state to now, I pulled out the `rss` specific logic and leveraged `document` logic where possible, adding handling by the `document_type` of either `rssfeed` or `rsspost` depending on the case. 

I've also simplified the scheduling/feed parsing/crawl queuing, which I updated in the PR description. 

Feature-wise, I've also removed the original hard-coded "follow links" and "link limit" parameters the RSS logic was using when sending a website for crawling. The user can now, in the UI, set follow links and limit, just like on the Add Website to Crawl page. 
On the RSS Feed details page, the user can toggle the RSS subscription on/off. The user can also modify the "follow links" and "link limit" values applied to the RSS feed, which will apply the settings to any newly found RSS feed posts.

The original requirement to bump the `aws-cdk-lib` version has been removed as I removed the `aws-solutions-construct` I was leveraging in the CDK stack. 

Within the `requirements.txt` there's an added dependency for `feedparser` https://pypi.org/project/feedparser/

massi-ang commented on lib/rag-engines/data-import/functions/trigger-rss-ingestors/index.py:
```diff
@@ -0,0 +1,21 @@
+    try:
+        genai_core.documents.ingest_rss_feeds()
+    except Exception as e:
+        logger.error(f'Error triggering RSS Feed checks')
```
remove `f` since the string is not a template

massi-ang commented on lib/rag-engines/data-import/functions/rss-ingestor/index.py:
```diff
@@ -0,0 +1,25 @@
+    try:
+        genai_core.documents.check_rss_feed_for_posts(workspace_id,document_id)
+    except Exception as e:
+        logger.error(f'Error checking for new posts from feed!')
```
remove `f` since the string is not a template

massi-ang commented on lib/rag-engines/data-import/website-crawling-workflow.ts:
Wouldn't be better to have a separate construct for the RSS Ingestor?

flamingquaks replied:
@massi-ang - This was something I wasn't 100% sure about. 
It's an extension of website crawling, but I can shift into an isolated construct if it would be better

massi-ang commented on lib/rag-engines/workspaces/delete-workspace.ts:
run `prettify` formatter

massi-ang commented on lib/shared/layers/python-sdk/python/genai_core/documents.py:
```diff
@@ -23,12 +27,19 @@
     "DEFAULT_KENDRA_S3_DATA_SOURCE_BUCKET_NAME"
 )
 
+RSS_FEED_INGESTOR_FUNCTION = os.environ['RSS_FEED_INGESTOR_FUNCTION'] if "RSS_FEED_INGESTOR_FUNCTION" in os.environ else ""
```
use `os.environ.get("VAR_NAME", "")`

massi-ang commented on lib/shared/layers/python-sdk/python/genai_core/documents.py:
```diff
@@ -23,12 +27,19 @@
 )
 
+RSS_FEED_INGESTOR_FUNCTION = os.environ['RSS_FEED_INGESTOR_FUNCTION'] if "RSS_FEED_INGESTOR_FUNCTION" in os.environ else ""
+RSS_FEED_SCHEDULE_ROLE_ARN = os.environ['RSS_FEED_SCHEDULE_ROLE_ARN'] if "RSS_FEED_SCHEDULE_ROLE_ARN" in os.environ else ""
```
same as above

massi-ang commented on lib/shared/layers/python-sdk/python/genai_core/documents.py:
```diff
@@ -48,6 +60,11 @@ def list_documents(
 
+    sort_key_prefix = f"{document_type}/"
+    if document_type == "rsspost" and parent_document_id != None:
+        sort_key_prefix = f"rsspost/{parent_document_id}/"
```
Can we make this more generic?
```
if parent_document_id != None:
  sort_key_prefix = sort_key_prefix+f"{parent_document_id}/"
```

massi-ang commented on lib/shared/layers/python-sdk/python/genai_core/documents.py:
run `black` formatter on the file

massi-ang commented on lib/shared/layers/python-sdk/python/genai_core/documents.py:
```diff
@@ -190,9 +208,30 @@ def set_status(workspace_id: str, document_id: str, status: str):
             ":timestampValue": timestamp,
         },
     )
+    update_timestamp(workspace_id, document_id)
```
the content of this function seems to be redundant with the above code. Was the idea to move the whole logic into the `update_timestamp` function?
the return value is not used, is this correct?

flamingquaks replied:
You are correct. I removed `update_timestamp` function. The `update_subscription_timestamp` still exists and is leveraged throughout the code.

massi-ang commented on lib/rag-engines/rag-dynamodb-tables/index.ts:
run `prettify` on this file

massi-ang commented on lib/shared/layers/python-sdk/python/genai_core/opensearch/delete.py:
revert this change

massi-ang commented on lib/shared/layers/python-sdk/python/genai_core/documents.py:
```diff
@@ -190,9 +208,30 @@ def set_status(workspace_id: str, document_id: str, status: str):
+    return response
 
+def update_timestamp(workspace_id: str, document_id: str):
+    timestamp = datetime.utcnow().strftime("%Y-%m-%dT%H:%M:%S.%fZ")
```
consider creating a function to return the DDB formatted timestamp since this is used in several places

massi-ang commented on lib/shared/layers/python-sdk/python/genai_core/documents.py:
```diff
@@ -500,3 +567,158 @@ def _upload_document_content(
+                    "errors": [],
+                    "created_at": timestamp,
+                    "updated_at": timestamp,
+                    "crawler_properties": rss_document['crawler_properties'] if 'crawler_properties' in rss_document else None,
```
replace with `dict.get("key", default)` expression

massi-ang commented on lib/shared/layers/python-sdk/python/genai_core/documents.py:
```diff
@@ -500,3 +567,158 @@ def _upload_document_content(
+                    "document_type": "rsspost",
+                    "document_sub_type": None,
+                    "sub_documents": None,
+                    "compound_sort_key": "rsspost/" + document_id + "/" + post_id,
```
maybe replace with `f"rsspost/{document_id}/{post_id}`

massi-ang commented on lib/user-interface/react-app/src/common/api-client/api-client.ts:
revert these changes

massi-ang commented on lib/user-interface/react-app/src/common/api-client/documents-client.ts:
run `prettify` on this file

massi-ang commented on lib/user-interface/react-app/src/common/types.ts:
```diff
@@ -137,8 +138,26 @@ export interface DocumentItem {
   subDocuments: number;
   createdAt: string;
   updatedAt: string;
+  rssFeedId?: string;
+  rssLastCheckedAt?: string;
+  crawlerProperties?: {
+    followLinks: boolean;
+    limit: number;
+  }
 }
```
can you explain why this information, which is related to the source of the document, must be included in the `DocumentItem` type?

flamingquaks replied:
On the `rss-feed.tsx` page, both the RSS Feed subscription and the RSS Feed posts display this information.
For the Feed, it's showing the currently configured crawler properties that will be leveraged by website crawling requests triggered by detecting new feed posts. 
On the individual feed records, it shows what was configured when the feed post was ingested and sent for crawling. 
![sample](https://github.com/aws-samples/aws-genai-llm-chatbot/assets/40297962/758ecfbe-f902-454a-bf80-2039f79c9d90)

massi-ang commented on lib/user-interface/react-app/src/pages/chatbot/playground/multi-chat-playground.tsx:
revert this change or run `prettify` on this file

massi-ang commented on lib/user-interface/react-app/src/pages/rag/workspace/documents-tab.tsx:
run `prettify` on this file

massi-ang commented on lib/user-interface/react-app/src/pages/rag/workspace/documents-tab.tsx:
```diff
@@ -101,12 +104,12 @@ export default function DocumentsTab(props: DocumentsTabProps) {
   const typeTitleStr = ragDocumentTypeToTitleString(props.documentType);
-
+  const columnDefinitions = getColumnDefinition(props.documentType);
   return (
```
add new line in between

massi-ang commented on lib/user-interface/react-app/src/pages/rag/workspace/rss-feed.tsx:
run `prettify` on this file

flamingquaks commented on lib/chatbot-api/functions/api-handler/routes/documents.py:
```diff
@@ -156,20 +229,73 @@ def add_document(workspace_id: str, document_type: str):
 
 
 def _convert_document(document: dict):
+    if "crawler_properties" in document:
```
@massi-ang I refactored this with the latest push to be a bit simpler. Basically the `crawler_properties` dict needs to format the objects within to have proper keys

flamingquaks commented:
@massi-ang, I moved RSS Subscription to its own construct.

flamingquaks commented on lib/shared/layers/python-sdk/python/genai_core/opensearch/delete.py:
@massi-ang - This appears in the PR as there was originally functionality in the workspace deletion, but has been removed. I don't know how to remove this blank change.
